text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/AnsiComponentSerializerImpl.java

```diff
@@ -0,0 +1,273 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.serializer.ansi;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.TranslatableComponent;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextColor;
+import net.kyori.adventure.text.format.TextDecoration;
+import net.kyori.adventure.text.format.TextFormat;
+import net.kyori.adventure.text.renderer.TranslatableComponentRenderer;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.fusesource.jansi.Ansi;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class AnsiComponentSerializerImpl implements AnsiComponentSerializer {
+
+  private final boolean colorDownSample;
+
+  @SuppressWarnings({"WeakerAccess"})
+  public static AnsiComponentSerializer trueColor = new AnsiComponentSerializerImpl(false);
+
+  private AnsiComponentSerializerImpl(final boolean colorDownSample) {
+    this.colorDownSample = colorDownSample;
+  }
+
+  private static final TextDecoration[] DECORATIONS = TextDecoration.values();
+  protected static final String RGB_FORMAT = ESC_CHAR + "[38;2;%d;%d;%dm";
+  private static final Pattern STRIP_ESC_CODES = Pattern.compile(ESC_CHAR + "\\[[;\\d]*m");
+  private static final List<TextFormat> FORMATS;
+
+  static {
+    final List<TextFormat> formats = new ArrayList<>(NamedTextColor.values());
+    Collections.addAll(formats, DECORATIONS);
+    formats.add(Reset.INSTANCE);
+    FORMATS = Collections.unmodifiableList(formats);
+
+  }
+
+  private Ansi toAnsi(final TextFormat format) {
+    if(format instanceof NamedTextColor) {
+      return Formats.byFormat(format).ansi();
+    }
+    if(format instanceof TextDecoration) {
+      return Formats.byFormat(format).ansi();
+    }
+    if(format instanceof TextColor) {
+      if(this.colorDownSample) {
+        return Formats.byFormat(NamedTextColor.nearestTo((TextColor) format)).ansi();
+      }
+      final int red = ((TextColor) format).red();
+      final int blue = ((TextColor) format).blue();
+      final int green = ((TextColor) format).green();
+      return Ansi.ansi().format(RGB_FORMAT, red, green, blue);
+    } else {
+      return null;
+    }
+  }
+
+  @NonNull
+  @Override
+  public Component deserialize(final @NonNull String input) {
```

Comment:
```suggestion
  @Override
  public @NonNull Component deserialize(final @NonNull String input) {
```

same for all your methods -- nullability annotations target the return type, not the method itself